[JSC] Further simplify Structure transitions

Constellation opened the pull request:
#### 33eeb781c9b5901be1d162e566d9f8bdc9bd7152
<pre>
[JSC] Further simplify Structure transitions
<a href="https://bugs.webkit.org/show_bug.cgi?id=256804">https://bugs.webkit.org/show_bug.cgi?id=256804</a>
rdar://109371189

Reviewed by Mark Lam.

Micro-optimize Put operation.

1. Inline Structure::pin.
2. Leverage overridesPut flag.

                                    ToT                     Patched

    dictionary-addition      111.8379+-0.6996     ^    106.8357+-0.6202        ^ definitely 1.0468x faster

* Source/JavaScriptCore/runtime/JSCellInlines.h:
(JSC::JSCell::putInline):
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::Structure::pin): Deleted.
* Source/JavaScriptCore/runtime/Structure.h:
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::pin):

Canonical link: <a href="https://commits.webkit.org/264106@main">https://commits.webkit.org/264106@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f9a355b4896462fc2e828fad07ba899dfec6b079

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6704 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6921 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7103 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8293 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6969 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8064 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6866 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9867 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6822 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7396 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6139 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8386 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/4842 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6067 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13890 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/5656 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/6610 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6145 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8873 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/6286 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6635 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5433 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/6841 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6027 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1594 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10200 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/7025 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/783 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6400 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1724 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/Structure.h:
```diff
@@ -988,7 +988,7 @@ class Structure : public JSCell {
     bool isValid(JSGlobalObject*, StructureChain* cachedPrototypeChain, JSObject* base) const;
 
     // You have to hold the structure lock to do these.
-    JS_EXPORT_PRIVATE void pin(const AbstractLocker&, VM&, PropertyTable*);
```
When `JS_EXPORT_PRIVATE` is attached, LTO even cannot inline it since this is explicitly exposed outside of this framework.

MenloDorian replied:
Maybe you should add a comment here so that no one mistakenly adds `JS_EXPORT_PRIVATE` back in the future.

MenloDorian reviewed (approved):
r=me

Constellation committed:
[JSC] Further simplify Structure transitions
https://bugs.webkit.org/show_bug.cgi?id=256804
rdar://109371189

Reviewed by Mark Lam.

Micro-optimize Put operation.

1. Inline Structure::pin.
2. Leverage overridesPut flag.

                                    ToT                     Patched

    dictionary-addition      111.8379+-0.6996     ^    106.8357+-0.6202        ^ definitely 1.0468x faster

* Source/JavaScriptCore/runtime/JSCellInlines.h:
(JSC::JSCell::putInline):
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::Structure::pin): Deleted.
* Source/JavaScriptCore/runtime/Structure.h:
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::pin):

Canonical link: https://commits.webkit.org/264106@main
